x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts

```diff
@@ -105,14 +131,43 @@ export class SecureSavedObjectsClientWrapper implements SavedObjectsClientContra
     attributes: Partial<T>,
     options: SavedObjectsUpdateOptions = {}
   ) {
-    await this.ensureAuthorized(type, 'update', options.namespace, {
-      type,
-      id,
-      attributes,
-      options,
+    const args = { type, id, attributes, options };
+    await this.ensureAuthorized(type, 'update', options.namespace, args);
+
+    const savedObject = await this.baseClient.update(type, id, attributes, options);
+    return await this.filterSavedObjectNamespaces(savedObject);
+  }
+
+  public async addNamespaces(
+    type: string,
+    id: string,
+    namespaces: string[],
+    options: SavedObjectsAddNamespacesOptions = {}
+  ) {
+    const args = { type, id, namespaces, options };
+    // to share an object, the user must have the "create" permission in all of the destination namespaces
+    await this.ensureAuthorized(type, 'create', namespaces, args, 'addNamespacesCreate');
+
+    return await this.baseClient.addNamespaces(type, id, namespaces, {
+      ...options,
+      validateExistingNamespaces: async (existing: string[]) => {
```

Comment:
I used this approach for a few reasons:
1. A pre-flight `get` (as I had in my prior implementation) would result in three cluster calls for this operation (`get`, `get` again, and `update`).
2. More importantly, a pre-flight `get` would result in information disclosure. It would allow an unauthorized user to make an API call to determine if a multi-namespace saved object exists in that space (404 response means it doesn't, 403 response means it does). All of our other operations in the `SecureSavedObjectsClientWrapper` don't behave this way, because none of them use any sort of pre-flight request. By using a hook/callback here I was able to sidestep this issue entirely.

The other approach I thought about to prevent information disclosure was to modify the Repository's `get` params to allow for a `ignoreNamespacesCheck` flag, which would change the behavior of its `get` cluster call:

https://github.com/elastic/kibana/blob/4f066029a042204e33b9ff2adff409aabcc6e8d9/src/core/server/saved_objects/service/lib/repository.ts#L880-L883

Reply:
> A pre-flight get (as I had in my prior implementation) would result in three cluster calls for this operation (get, get again, and update). This approach only uses two cluster calls.

I do appreciate trying to optimize the number of requests here. This will become even more important as we build out audit logging. My personal preference for this endpoint is to revert to the pre-flight check (assuming my assumption around information disclosure below holds). I don't expect that the `addNamespaces` endpoint will be part of any "hot" code paths. In other words, an extra round-trip for this operation isn't likely to be noticed. If this were part of a more common operation, then it might be worthwhile to optimize this, but I don't _personally_ think the extra complexity and core service hook justify the gains here.


> More importantly, a pre-flight get would result in information disclosure.

If we had the pre-flight `get` call within a try/catch block here, couldn't we control the response the end-user receives?

Reply:
> If we had the pre-flight `get` call within a try/catch block here, couldn't we control the response the end-user receives?
Well, here's the scenario that I'm thinking of.

* Object `sharedtype:foo` exists in `space-a` and `space-b`
* User calls the API endpoint to add namespaces within Space C (e.g., `/s/space-c/api/spaces/_share_saved_object_add`

The `SecureSavedObjectsClientWrapper` will do its pre-flight `get` call and receive a 404 result, because that call is in the context of `space-c`. Now, we don't know if that is actually because the object doesn't exist at all, or because the object doesn't exist in this space.

We could catch any 404 error and return a 403 instead, but I don't like that idea because then we'll receive different results depending on which spaceid-prefix we use for the API call. 

This is just a corner case, but I am hesitant to introduce unpredictable behavior like that.

Alternatively, we could always make `addNamespaces` and `removeNamespaces` calls scoped in the context of the current namespace. That would at least ensure that behavior of both is consistent (you'd get the 404 result for both, given the scenario above).

Reply:
> > If we had the pre-flight `get` call within a try/catch block here, couldn't we control the response the end-user receives?
> > Well, here's the scenario that I'm thinking of.
> 
> * Object `sharedtype:foo` exists in `space-a` and `space-b`
> * User calls the API endpoint to add namespaces within Space C (e.g., `/s/space-c/api/spaces/_share_saved_object_add`
> 
> The `SecureSavedObjectsClientWrapper` will do its pre-flight `get` call and receive a 404 result, because that call is in the context of `space-c`. Now, we don't know if that is actually because the object doesn't exist at all, or because the object doesn't exist in this space.
> 
> We could catch any 404 error and return a 403 instead, but I don't like that idea because then we'll receive different results depending on which spaceid-prefix we use for the API call.
> 
> This is just a corner case, but I am hesitant to introduce unpredictable behavior like that.
> 
> Alternatively, we could always make `addNamespaces` and `removeNamespaces` calls scoped in the context of the current namespace. That would at least ensure that behavior of both is consistent (you'd get the 404 result for both, given the scenario above).

I like your alternative proposal of scoping these calls to the context of the current namespace. That's actually how I'd expect them to work, as that would be consistent with the rest of the saved objects client.